src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -674,6 +675,76 @@ ss::future<init_tm_tx_reply> tx_gateway_frontend::do_init_tm_tx(
       });
 }
 
+namespace {
+
+bool is_epoch_overflow(int16_t epoch) {
+    return epoch >= std::numeric_limits<int16_t>::max();
+}
+
+bool is_valid_expected_pid(
+  const tm_transaction& tx, const model::producer_identity& expected_pid) {
+    if (expected_pid == model::unknow_pid) {
+        return true;
+    }
+
+    return tx.pid.id == expected_pid.id || ((tx.last_pid.id == expected_pid.id) && is_epoch_overflow(expected_pid.epoch));
+}
+
+bool is_valid_overflow(
+  const tm_transaction& tx, model::producer_identity expected_pid) {
+    if (!is_epoch_overflow(tx.pid.epoch)) {
+        return false;
+    }
+
+    if (expected_pid == model::unknow_pid) {
+        return true;
+    }
+
+    return tx.pid.epoch == expected_pid.epoch;
+}
+
+using current_and_last_epoch
```

Comment:
I got a review comment else where that std::pair has a readability problem (first, second etc), rather switch to a struct.